From: Marc Kleine-Budde <mkl@pengutronix.de>
Date: Tue, 20 Oct 2015 10:29:04 +0200
Subject: [PATCH] of_path: add of_find_path_by_node()

This patch adds the function of_find_path_by_node(), which is similar to
of_find_path(), but it translates a device tree node into a barebox device path
directly.

Signed-off-by: Marc Kleine-Budde <mkl@pengutronix.de>
---
 drivers/of/of_path.c | 20 +++++++++++++++++++-
 include/of.h         |  1 +
 2 files changed, 20 insertions(+), 1 deletion(-)

diff --git a/drivers/of/of_path.c b/drivers/of/of_path.c
index 65fb31bcec59..0e0b43e3a1b2 100644
--- a/drivers/of/of_path.c
+++ b/drivers/of/of_path.c
@@ -113,6 +113,9 @@ static int __of_find_path(struct device_node *node, struct device_node *rnode, c
 	bool add_bb = false;
 	int i, ret;
 
+	if (!rnode)
+		rnode = node;
+
 	op.dev = of_find_device_by_node_path(rnode->full_name);
 	if (!op.dev) {
 		op.dev = of_find_device_by_node_path(rnode->parent->full_name);
@@ -126,7 +129,7 @@ static int __of_find_path(struct device_node *node, struct device_node *rnode, c
 
 	i = 1;
 
-	while (1) {
+	while (propname) {
 		ret = of_property_read_string_index(node, propname, i++, &str);
 		if (ret)
 			break;
@@ -149,6 +152,21 @@ static int __of_find_path(struct device_node *node, struct device_node *rnode, c
 }
 
 /**
+ * of_find_path_by_node - translate a node in the devicetree to a
+ *                     	  barebox device path
+ *
+ * @node: the node we're interested in
+ * @outpath: if this function returns 0 outpath will contain the path belonging
+ *           to the input path description. Must be freed with free().
+ * @flags: use OF_FIND_PATH_FLAGS_BB to return the .bb device if available
+ *
+ */
+int of_find_path_by_node(struct device_node *node, char **outpath, unsigned flags)
+{
+	return __of_find_path(node, NULL, NULL, outpath, flags);
+}
+
+/**
  * of_find_path - translate a path description in the devicetree to a barebox
  *                path
  *
diff --git a/include/of.h b/include/of.h
index e0ebc39b742d..e60fe8982504 100644
--- a/include/of.h
+++ b/include/of.h
@@ -246,6 +246,7 @@ void of_add_memory_bank(struct device_node *node, bool dump, int r,
 struct device_d *of_find_device_by_node_path(const char *path);
 #define OF_FIND_PATH_FLAGS_BB 1		/* return .bb device if available */
 int of_find_path(struct device_node *node, const char *propname, char **outpath, unsigned flags);
+int of_find_path_by_node(struct device_node *node, char **outpath, unsigned flags);
 int of_register_fixup(int (*fixup)(struct device_node *, void *), void *context);
 int of_unregister_fixup(int (*fixup)(struct device_node *, void *), void *context);
 struct device_node *of_find_node_by_alias(struct device_node *root,
